Analyzer tests and fixes for analyzer/additionalfiles/rulesets

JurjenBiewenga opened the pull request.

JurjenBiewenga committed:
Added tests for analyzers/additional files/rulesets

Fixed analyzers and additional files delimited by `;`

van800 commented on Packages/com.unity.ide.rider/Rider/Editor/ProjectGeneration/ProjectGeneration.cs:
```diff
@@ -764,6 +764,7 @@ static string GetProjectHeaderTemplate()
         @"    <_TargetFrameworkDirectories>non_empty_path_generated_by_unity.rider.package</_TargetFrameworkDirectories>",
         @"    <_FullFrameworkReferenceAssemblyPaths>non_empty_path_generated_by_unity.rider.package</_FullFrameworkReferenceAssemblyPaths>", 
         @"    <DisableHandlePackageFileConflicts>true</DisableHandlePackageFileConflicts>",
+        @"{16}",
```
this would produce empty line in csproj

JurjenBiewenga replied:
Good catch

van800 commented on Packages/com.unity.ide.rider.tests/Tests/Editor/CSProjectTests.cs:
```diff
@@ -216,24 +217,79 @@ public void AllowUnsafeBlock()
             }
+
+            [TestCase(new Object[] {"C:/Analyzer.dll"})]
+            [TestCase(new Object[] {"C:/Analyzer.dll", "C:/Analyzer2.dll"})]
```
Passing actual content of csc.rsp would be more flexible. It may be:
In one case: 
```
-a:C:\Work\com.unity.ide.rider\Test\Data\ProjectGenerationTests\RoslynAnalyzers\analyzers\dotnet\cs\Comments_analyser.dll;C:\Work\com.unity.ide.rider\Test\Data\ProjectGenerationTests\RoslynAnalyzers\analyzers\dotnet\cs\UnityEngineAnalyzer.dll
-a:Comments_analyser1.dll
```
In second case just an empty string.
In 3rd case just:
`-a:Comments_analyser1.dll`

JurjenBiewenga replied:
Wouldn't we rather have separate tests for the parser?

van800 replied:
Ok, fair enough. Then just use `new string[]` instead of `new Object[]`.

JurjenBiewenga replied:
I tried to use `new string[]` however NUnit didn't like it, perhaps there's some other way. I'll try again.

van800 replied:
Oh, sorry. I see you can only do `object` instead of `Object` there.

van800 commented on Packages/com.unity.ide.rider/Rider/Editor/ProjectGeneration/ProjectGeneration.cs:
```diff
@@ -688,8 +688,8 @@ List<ResponseFileData> responseFilesData
         GenerateNoWarn(otherResponseFilesData["nowarn"].ToArray()),
-        GenerateAnalyserItemGroup(otherResponseFilesData["analyzer"].Concat(otherResponseFilesData["a"]).Distinct().ToArray()),
-        GenerateAnalyserAdditionalFiles(otherResponseFilesData["additionalfile"].ToArray()),
+        GenerateAnalyserItemGroup(otherResponseFilesData["analyzer"].Concat(otherResponseFilesData["a"]).Distinct().SelectMany(x=>x.Split(';')).ToArray()),
```
I guess `Distinct` should go after splitting.

JurjenBiewenga replied:
Will fix, thanks!

JurjenBiewenga committed:
Fixed empty lines in csprojectgeneration

Replace Object with the type keyword
Added testcase for multiple ruleset definitions

JurjenBiewenga commented:
Currently multiple rulesets are supported, I'm not sure if this actually is allowed by the spec. @van800 Do you happen to know? If not should we let the user know?

van800 commented:
From what I see, only last ruleset is applied.

Tried that with analyser from:
`https://github.com/vad710/UnityEngineAnalyzer/releases`

csc.rsp
```
-a:"C:\Work\com.unity.ide.rider\Test\Data\ProjectGenerationTests\RoslynAnalyzers\analyzers\dotnet\cs\UnityEngineAnalyzer.dll" -ruleset:"C:\Work\com.unity.ide.rider\Test\Data\ProjectGenerationTests\rules2.ruleset" -ruleset:"C:\Work\com.unity.ide.rider\Test\Data\ProjectGenerationTests\rules.ruleset"
```

rules.ruleset
```
<?xml version="1.0" encoding="utf-8"?> <RuleSet Name="My Rules" Description="Customized." ToolsVersion="10.0">   <Rules AnalyzerId="UnityEngineAnalyzer" RuleNamespace="UnityEngineAnalyzer">     <Rule Id="AOT0001" Action="Error" />   </Rules> </RuleSet>
```

van800 commented on Packages/com.unity.ide.rider.tests/Tests/Editor/CSProjectTests.cs:
```diff
@@ -6,6 +6,7 @@
 using NUnit.Framework;
 using UnityEditor.Compilation;
 using UnityEngine;
+using Object = System.Object;
```
Cleanup unused usings.

van800 commented:
I think we should leave current behaviour as is.
With multiple rulesets both Unity and Rider use only last supplied ruleset without throwing any error.

van800 commented on Packages/com.unity.ide.rider/Rider/Editor/ProjectGeneration/ProjectGeneration.cs:
```diff
@@ -895,7 +895,7 @@ private static string GenerateAnalyserRuleSet(string[] paths)
             return string.Empty;
       
-        return string.Join("\r\n", paths.Select(a => $"\r\n  <CodeAnalysisRuleSet>{a}</CodeAnalysisRuleSet>"));
+        return $"{Environment.NewLine}{string.Join(Environment.NewLine, paths.Select(a => $"  <CodeAnalysisRuleSet>{a}</CodeAnalysisRuleSet>"))}";
```
I have asked around and it seems those "\r\n" instead of Environment.NewLine are there for a reason.
Could you please revert them back.

JurjenBiewenga replied:
Will do

JurjenBiewenga committed:
Removed unused usings

JurjenBiewenga committed:
Replaced Environment.NewLine with "\r\n"

van800 commented on Packages/com.unity.ide.rider/Rider/Editor/ProjectGeneration/ProjectGeneration.cs:
```diff
@@ -895,7 +895,7 @@ private static string GenerateAnalyserRuleSet(string[] paths)
             return string.Empty;
       
-        return $"{Environment.NewLine}{string.Join(Environment.NewLine, paths.Select(a => $"  <CodeAnalysisRuleSet>{a}</CodeAnalysisRuleSet>"))}";
+        return $"\r\n{string.Join("\r\n", paths.Select(a => $"  <CodeAnalysisRuleSet>{a}</CodeAnalysisRuleSet>"))}";
```
Would produce "\r\n" for empty list, right?
